File: src/Python/devpossible_ton/devpossible_ton/validator/ton_validator.py
"""
TonValidator - Validator for TON format
Copyright (c) 2024 DevPossible, LLC
"""

from dataclasses import dataclass, field
from typing import Dict, List, Any, Optional
from ..models import TonDocument, TonObject, TonValue, TonArray
from ..errors import TonValidationError


@dataclass
class TonValidationResult:
    """Result of validation."""
    is_valid: bool
    errors: List[str] = field(default_factory=list)
    warnings: List[str] = field(default_factory=list)


class TonValidator:
    """Validator for TON format."""

    def validate(self, document: TonDocument, schema: Dict[str, Any]) -> TonValidationResult:
        """Validate a document against a schema."""
        result = TonValidationResult(is_valid=True)

        try:
            self._validate_value(document.get_root(), schema, '', result)
        except TonValidationError as e:
            result.is_valid = False
            result.errors.append(str(e))

        return result

    def _validate_value(self, value: Any, schema: Dict[str, Any], path: str,
                        result: TonValidationResult) -> None:
        """Validate a value against a schema."""
        schema_type = schema.get('type')

        if schema_type == 'object':
            self._validate_object(value, schema, path, result)
        elif schema_type == 'array':
            self._validate_array(value, schema, path, result)
        elif schema_type == 'string':
            self._validate_string(value, schema, path, result)
        elif schema_type == 'number':
            self._validate_number(value, schema, path, result)
        elif schema_type == 'boolean':
            self._validate_boolean(value, schema, path, result)
        elif schema_type == 'null':
            self._validate_null(value, schema, path, result)

    def _validate_object(self, value: Any, schema: Dict[str, Any], path: str,
                         result: TonValidationResult) -> None:
        """Validate an object."""
        if not isinstance(value, TonObject):
            result.is_valid = False
            result.errors.append(f"{path}: Expected object, got {type(value).__name__}")
            return

        # Check required properties
        required = schema.get('required', [])
        for req_prop in required:
            if not value.has(req_prop):
                result.is_valid = False
                result.errors.append(f"{path}: Missing required property '{req_prop}'")

        # Validate properties
        properties_schema = schema.get('properties', {})
        for key in value.keys():
            if key in properties_schema:
                prop_path = f"{path}.{key}" if path else key
                prop_value = value.get(key)
                self._validate_value(prop_value, properties_schema[key], prop_path, result)

    def _validate_array(self, value: Any, schema: Dict[str, Any], path: str,
                        result: TonValidationResult) -> None:
        """Validate an array."""
        if not isinstance(value, TonArray):
            result.is_valid = False
            result.errors.append(f"{path}: Expected array, got {type(value).__name__}")
            return

        # Check min/max items
        min_items = schema.get('minItems')
        max_items = schema.get('maxItems')
        length = value.length()

        if min_items is not None and length < min_items:
            result.is_valid = False
            result.errors.append(f"{path}: Array has {length} items, minimum is {min_items}")

        if max_items is not None and length > max_items:
            result.is_valid = False
            result.errors.append(f"{path}: Array has {length} items, maximum is {max_items}")

        # Validate items
        items_schema = schema.get('items')
        if items_schema:
            for i in range(length):
                item_path = f"{path}[{i}]"
                item_value = value.get(i)
                self._validate_value(item_value, items_schema, item_path, result)

    def _validate_string(self, value: Any, schema: Dict[str, Any], path: str,
                         result: TonValidationResult) -> None:
        """Validate a string."""
        actual_value = value.get_value() if isinstance(value, TonValue) else value

        if not isinstance(actual_value, str):
            result.is_valid = False
            result.errors.append(f"{path}: Expected string, got {type(actual_value).__name__}")
            return

        # Check enum constraint
        enum_values = schema.get('enum')
        if enum_values is not None and actual_value not in enum_values:
            result.is_valid = False
            result.errors.append(f'{path}: Value "{actual_value}" is not in enum {enum_values}')
            return

        # Check min/max length
        min_length = schema.get('minLength')
        max_length = schema.get('maxLength')

        if min_length is not None and len(actual_value) < min_length:
            result.is_valid = False
            result.errors.append(f"{path}: String length {len(actual_value)} is less than minimum {min_length}")

        if max_length is not None and len(actual_value) > max_length:
            result.is_valid = False
            result.errors.append(f"{path}: String length {len(actual_value)} exceeds maximum {max_length}")

    def _validate_number(self, value: Any, schema: Dict[str, Any], path: str,
                         result: TonValidationResult) -> None:
        """Validate a number."""
        actual_value = value.get_value() if isinstance(value, TonValue) else value

        if not isinstance(actual_value, (int, float)):
            result.is_valid = False
            result.errors.append(f"{path}: Expected number, got {type(actual_value).__name__}")
            return

        # Check min/max
        minimum = schema.get('minimum')
        maximum = schema.get('maximum')

        if minimum is not None and actual_value < minimum:
            result.is_valid = False
            result.errors.append(f"{path}: Value {actual_value} is less than minimum {minimum}")

        if maximum is not None and actual_value > maximum:
            result.is_valid = False
            result.errors.append(f"{path}: Value {actual_value} exceeds maximum {maximum}")

    def _validate_boolean(self, value: Any, schema: Dict[str, Any], path: str,
                          result: TonValidationResult) -> None:
        """Validate a boolean."""
        actual_value = value.get_value() if isinstance(value, TonValue) else value

        if not isinstance(actual_value, bool):
            result.is_valid = False
            result.errors.append(f"{path}: Expected boolean, got {type(actual_value).__name__}")

    def _validate_null(self, value: Any, schema: Dict[str, Any], path: str,
                       result: TonValidationResult) -> None:
        """Validate null."""
        actual_value = value.get_value() if isinstance(value, TonValue) else value

        if actual_value is not None:
            result.is_valid = False
            result.errors.append(f"{path}: Expected null, got {type(actual_value).__name__}")

File: src/CSharp/DevPossible.Ton/src/Validator/TonValidator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace DevPossible.Ton
{
    /// <summary>
    /// Validator for TON documents and objects
    /// </summary>
    public class TonValidator
    {
        private readonly TonSchemaCollection? _schemas;

        /// <summary>
        /// Creates a new validator with the given schemas
        /// </summary>
        public TonValidator(TonSchemaCollection? schemas = null)
        {
            _schemas = schemas;
        }

        /// <summary>
        /// Validates a TON document
        /// </summary>
        public TonValidationResult Validate(TonDocument document)
        {
            var result = new TonValidationResult { IsValid = true };

            if (_schemas == null && document.Schemas == null)
            {
                // No schemas to validate against
                return result;
            }

            var schemas = _schemas ?? document.Schemas;
            ValidateObject(document.RootObject, schemas!, result, "/");

            return result;
        }

        /// <summary>
        /// Validates a TON object
        /// </summary>
        public TonValidationResult ValidateObject(TonObject obj, TonSchemaCollection schemas)
        {
            var result = new TonValidationResult { IsValid = true };
            ValidateObject(obj, schemas, result, "/");
            return result;
        }

        private void ValidateObject(TonObject obj, TonSchemaCollection schemas, TonValidationResult result, string path)
        {
            if (obj.ClassName != null)
            {
                var schema = schemas.GetSchema(obj.ClassName);
                if (schema != null)
                {
                    ValidateAgainstSchema(obj, schema, schemas, result, path);
                }
            }

            // Validate child objects recursively
            foreach (var child in obj.Children)
            {
                var childPath = path + (child.ClassName ?? "object") + "/";
                ValidateObject(child, schemas, result, childPath);
            }
        }

        private void ValidateAgainstSchema(TonObject obj, TonSchemaDefinition schema, TonSchemaCollection schemas,
            TonValidationResult result, string path)
        {
            // Validate all properties defined in the schema
            foreach (var propSchema in schema.Properties.Values)
            {
                var schemaPath = propSchema.Path;
                var value = GetValueAtPath(obj, schemaPath);
                var fullPath = path + schemaPath.TrimStart('/');

                // Check if required property is missing
                if (propSchema.IsRequired && (value == null || value.IsUndefined))
                {
                    // Use the simple property name for backwards compatibility with existing tests
                    var simpleName = schemaPath.TrimStart('/').Split('/').LastOrDefault() ?? schemaPath;
                    result.AddError($"Required property '{simpleName}' is missing", fullPath);
                    continue;
                }

                // Validate property if it exists
                if (value != null && !value.IsUndefined)
                {
                    ValidateProperty(value, propSchema, schemas, result, fullPath);
                }
            }
        }

        private TonValue? GetValueAtPath(TonObject obj, string path)
        {
            // Remove leading slash if present
            path = path.TrimStart('/');

            // Split path into segments
            var segments = path.Split('/');
            TonObject? current = obj;
            TonValue? value = null;

            for (int i = 0; i < segments.Length; i++)
            {
                var segment = segments[i];

                if (current == null)
                    return null;

                // Get property from current object
                value = current.GetProperty(segment);

                // If this is not the last segment, we need to traverse deeper
                if (i < segments.Length - 1)
                {
                    // Check if value is an object we can traverse into
                    if (value?.Value is TonObject nestedObj)
                    {
                        current = nestedObj;
                    }
                    else
                    {
                        // Can't traverse deeper - path doesn't exist
                        return null;
                    }
                }
            }

            return value;
        }

        private void ValidateProperty(TonValue value, TonPropertySchema schema, TonSchemaCollection schemas,
            TonValidationResult result, string path)
        {
            // Check NotNull validation first if present
            var hasNotNullRule = schema.Validations.Any(r => r.Type == ValidationRuleType.NotNull);

            // Type validation - allow null if NotNull validation will check it
            if (!ValidateType(value, schema.Type, schemas, allowNullForNotNullRule: hasNotNullRule))
            {
                // Only report type mismatch if it's not a null issue that NotNull will handle
                if (!(value.Type == TonValueType.Null && hasNotNullRule))
                {
                    result.AddError($"Type mismatch: expected {schema.Type}, got {value.Type}", path);
                    return;
                }
            }

            // Apply validation rules
            foreach (var rule in schema.Validations)
            {
                ValidateRule(value, rule, result, path);
            }
        }

        private bool ValidateType(TonValue value, string expectedType, TonSchemaCollection schemas, bool allowNullForNotNullRule = false)
        {
            // Allow null values if NotNull rule will validate them
            if (allowNullForNotNullRule && value.Type == TonValueType.Null)
                return true;

            switch (expectedType.ToLowerInvariant())
            {
                case "string":
                    return value.Type == TonValueType.String;
                case "int":
                case "integer":
                    return value.Type == TonValueType.Integer;
                case "float":
                case "double":
                case "decimal":
                    return value.Type == TonValueType.Float || value.Type == TonValueType.Integer;
                case "boolean":
                case "bool":
                    return value.Type == TonValueType.Boolean;
                case "null":
                    return value.Type == TonValueType.Null;
                case "undefined":
                    return value.Type == TonValueType.Undefined;
                case "guid":
                case "uuid":
                    return value.Type == TonValueType.Guid;
                case "date":
                case "datetime":
                    return value.Type == TonValueType.Date;
                case "array":
                    return value.Type == TonValueType.Array;
                default:
                    // Check for array with base type (array:baseType)
                    if (expectedType.StartsWith("array:"))
                    {
                        if (value.Type != TonValueType.Array)
                            return false;

                        // Validate base type for all elements
                        var baseType = expectedType.Substring(6);
                        var array = value.ToArray();
                        if (array != null)
                        {
                            foreach (var element in array)
                            {
                                if (!ValidateType(element, baseType, schemas))
                                    return false;
                            }
                        }
                        return true;
                    }
                    // Check if it's an enum type
                    if (expectedType.StartsWith("enum:"))
                    {
                        var enumName = expectedType.Substring(5);
                        var enumDef = schemas.GetEnum(enumName);
                        if (enumDef != null && !enumDef.IsEnumSet)
                        {
                            return value.Type == TonValueType.Enum && ValidateEnumValue(value, enumDef);
                        }
                    }
                    else if (expectedType.StartsWith("enumSet:"))
                    {
                        var enumName = expectedType.Substring(8);
                        var enumDef = schemas.GetEnum(enumName);
                        if (enumDef != null && enumDef.IsEnumSet)
                        {
                            return value.Type == TonValueType.EnumSet && ValidateEnumSetValue(value, enumDef);
                        }
                    }
                    return true; // Unknown type, assume valid
            }
        }

        private bool ValidateEnumValue(TonValue value, TonEnumDefinition enumDef)
        {
            if (value.Value is TonEnum enumValue)
            {
                return enumDef.IsValidValue(enumValue.Value);
            }
            return false;
        }

        private bool ValidateEnumSetValue(TonValue value, TonEnumDefinition enumDef)
        {
            if (value.Value is TonEnumSet enumSet)
            {
                return enumSet.Values.All(v => enumDef.IsValidValue(v.Value));
            }
            return false;
        }

        private void ValidateRule(TonValue value, TonValidationRule rule, TonValidationResult result, string path)
        {
            switch (rule.Type)
            {
                case ValidationRuleType.Required:
                    if (value.IsUndefined)
                        result.AddError("Value is required", path);
                    break;

                case ValidationRuleType.NotNull:
                    if (value.IsNull)
                        result.AddError("Value cannot be null", path);
                    break;

                case ValidationRuleType.MinLength:
                    if (value.Type == TonValueType.String && rule.Parameters?.Length > 0)
                    {
                        var minLength = Convert.ToInt32(rule.Parameters[0]);
                        var str = value.ToString() ?? "";
                        if (str.Length < minLength)
                            result.AddError($"String length must be at least {minLength}", path);
                    }
                    break;

                case ValidationRuleType.MaxLength:
                    if (value.Type == TonValueType.String && rule.Parameters?.Length > 0)
                    {
                        var maxLength = Convert.ToInt32(rule.Parameters[0]);
                        var str = value.ToString() ?? "";
                        if (str.Length > maxLength)
                            result.AddError($"String length must be at most {maxLength}", path);
                    }
                    break;

                case ValidationRuleType.Min:
                    if ((value.Type == TonValueType.Integer || value.Type == TonValueType.Float) && rule.Parameters?.Length > 0)
                    {
                        var min = Convert.ToDouble(rule.Parameters[0]);
                        if (value.ToDouble() < min)
                            result.AddError($"Value must be at least {min}", path);
                    }
                    break;

                case ValidationRuleType.Max:
                    if ((value.Type == TonValueType.Integer || value.Type == TonValueType.Float) && rule.Parameters?.Length > 0)
                    {
                        var max = Convert.ToDouble(rule.Parameters[0]);
                        if (value.ToDouble() > max)
                            result.AddError($"Value must be at most {max}", path);
                    }
                    break;

                case ValidationRuleType.Pattern:
                    if (value.Type == TonValueType.String && rule.Parameters?.Length > 0)
                    {
                        var pattern = rule.Parameters[0]?.ToString();
                        if (!string.IsNullOrEmpty(pattern))
                        {
                            var regex = new Regex(pattern);
                            var str = value.ToString() ?? "";
                            if (!regex.IsMatch(str))
                                result.AddError($"Value does not match pattern: {pattern}", path);
                        }
                    }
                    break;

                // Array-specific validations
                case ValidationRuleType.MinCount:
                    if (value.Type == TonValueType.Array && rule.Parameters?.Length > 0)
                    {
                        var minCount = Convert.ToInt32(rule.Parameters[0]);
                        var count = value.GetArrayCount();
                        if (count < minCount)
                            result.AddError($"Array must have at least {minCount} elements", path);
                    }
                    break;

                case ValidationRuleType.MaxCount:
                    if (value.Type == TonValueType.Array && rule.Parameters?.Length > 0)
                    {
                        var maxCount = Convert.ToInt32(rule.Parameters[0]);
                        var count = value.GetArrayCount();
                        if (count > maxCount)
                            result.AddError($"Array must have at most {maxCount} elements", path);
                    }
                    break;

                case ValidationRuleType.Count:
                    if (value.Type == TonValueType.Array && rule.Parameters?.Length > 0)
                    {
                        var expectedCount = Convert.ToInt32(rule.Parameters[0]);
                        var count = value.GetArrayCount();
                        if (count != expectedCount)
                            result.AddError($"Array must have exactly {expectedCount} elements", path);
                    }
                    break;

                case ValidationRuleType.NonEmpty:
                    if (value.Type == TonValueType.Array)
                    {
                        if (value.GetArrayCount() == 0)
                            result.AddError("Array must not be empty", path);
                    }
                    break;

                case ValidationRuleType.Unique:
                    if (value.Type == TonValueType.Array)
                    {
                        var array = value.ToArray();
                        if (array != null)
                        {
                            var seen = new HashSet<string>();
                            for (int i = 0; i < array.Count; i++)
                            {
                                var elementStr = array[i].ToString() ?? "null";
                                if (!seen.Add(elementStr))
                                {
                                    result.AddError($"Array must contain unique elements, duplicate found: {elementStr}", path);
                                    break;
                                }
                            }
                        }
                    }
                    break;

                case ValidationRuleType.Sorted:
                    if (value.Type == TonValueType.Array)
                    {
                        var array = value.ToArray();
                        if (array != null && array.Count > 1)
                        {
                            for (int i = 1; i < array.Count; i++)
                            {
                                var prev = array[i - 1];
                                var curr = array[i];

                                // Compare based on type
                                bool outOfOrder = false;
                                if (prev.Type == TonValueType.Integer && curr.Type == TonValueType.Integer)
                                {
                                    outOfOrder = prev.ToInt64() > curr.ToInt64();
                                }
                                else if ((prev.Type == TonValueType.Float || prev.Type == TonValueType.Integer) &&
                                         (curr.Type == TonValueType.Float || curr.Type == TonValueType.Integer))
                                {
                                    outOfOrder = prev.ToDouble() > curr.ToDouble();
                                }
                                else if (prev.Type == TonValueType.String && curr.Type == TonValueType.String)
                                {
                                    outOfOrder = string.Compare(prev.ToString(), curr.ToString(), StringComparison.Ordinal) > 0;
                                }

                                if (outOfOrder)
                                {
                                    result.AddError($"Array must be sorted, element at index {i} is out of order", path);
                                    break;
                                }
                            }
                        }
                    }
                    break;

                case ValidationRuleType.AllowDuplicates:
                    // This is a positive validation, no check needed (duplicates are allowed by default)
                    break;

                // Add more validation rule implementations as needed
            }
        }
    }

    /// <summary>
    /// Result of TON validation
    /// </summary>
    public class TonValidationResult
    {
        /// <summary>
        /// Whether the validation passed
        /// </summary>
        public bool IsValid { get; set; }

        /// <summary>
        /// List of validation errors
        /// </summary>
        public List<TonValidationError> Errors { get; } = new List<TonValidationError>();

        /// <summary>
        /// List of validation warnings
        /// </summary>
        public List<TonValidationWarning> Warnings { get; } = new List<TonValidationWarning>();

        /// <summary>
        /// Adds a validation error
        /// </summary>
        public void AddError(string message, string path)
        {
            IsValid = false;
            Errors.Add(new TonValidationError(message, path));
        }

        /// <summary>
        /// Adds a validation warning
        /// </summary>
        public void AddWarning(string message, string path)
        {
            Warnings.Add(new TonValidationWarning(message, path));
        }
    }

    /// <summary>
    /// Represents a validation error
    /// </summary>
    public class TonValidationError
    {
        /// <summary>
        /// Error message
        /// </summary>
        public string Message { get; }

        /// <summary>
        /// Path to the property that failed validation
        /// </summary>
        public string Path { get; }

        public TonValidationError(string message, string path)
        {
            Message = message;
            Path = path;
        }

        public override string ToString() => $"{Path}: {Message}";
    }

    /// <summary>
    /// Represents a validation warning
    /// </summary>
    public class TonValidationWarning
    {
        /// <summary>
        /// Warning message
        /// </summary>
        public string Message { get; }

        /// <summary>
        /// Path to the property that triggered the warning
        /// </summary>
        public string Path { get; }

        public TonValidationWarning(string message, string path)
        {
            Message = message;
            Path = path;
        }

        public override string ToString() => $"{Path}: {Message}";
    }
}